File: src/main/java/com/tobe/healthy/lessonhistory/domain/LessonHistoryComment.java
package com.tobe.healthy.lessonhistory.domain;

import static jakarta.persistence.CascadeType.*;
import static jakarta.persistence.FetchType.*;
import static jakarta.persistence.GenerationType.*;

import java.util.ArrayList;
import java.util.List;

import org.hibernate.annotations.ColumnDefault;
import org.hibernate.annotations.DynamicUpdate;

import com.tobe.healthy.common.BaseTimeEntity;
import com.tobe.healthy.member.domain.Member;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Transient;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;

@Entity
@DynamicUpdate
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@ToString
public class LessonHistoryComment extends BaseTimeEntity<LessonHistoryComment, Long> {

	@Id
	@GeneratedValue(strategy = IDENTITY)
	@Column(name = "lesson_history_comment_id")
	private Long id;

	@ManyToOne(fetch = LAZY)
	@JoinColumn(name = "parent_id")
	@ToString.Exclude
	private LessonHistoryComment parent;

	@Column(name = "\"order\"")
	private int order;

	private String content;

	@ManyToOne(fetch = LAZY)
	@JoinColumn(name = "writer_id")
	@ToString.Exclude
	private Member writer;

	@ManyToOne(fetch = LAZY)
	@JoinColumn(name = "lesson_history_id")
	@ToString.Exclude
	private LessonHistory lessonHistory;

	@OneToMany(fetch = LAZY, mappedBy = "lessonHistoryComment", cascade = ALL)
	@ToString.Exclude
	private List<LessonHistoryFiles> files = new ArrayList<>();

	@ColumnDefault("false")
	private boolean delYn = false;

	@Transient
	private List<LessonHistoryComment> replies = new ArrayList<>();

	public LessonHistoryComment(int order, String content, Member writer, LessonHistory lessonHistory) {
		this.order = order;
		this.content = content;
		this.writer = writer;
		this.lessonHistory = lessonHistory;
	}

	public LessonHistoryComment(int order, String content, Member writer, LessonHistory lessonHistory,
		LessonHistoryComment parent) {
		this.order = order;
		this.content = content;
		this.writer = writer;
		this.lessonHistory = lessonHistory;
		this.parent = parent;
	}

	public void updateLessonHistoryComment(String content) {
		this.content = content;
	}

	public void deleteComment() {
		this.delYn = true;
	}

	public void setReplies(List<LessonHistoryComment> replies) {
		this.replies = replies;
	}
}

File: src/main/kotlin/com/tobe/healthy/lessonhistory/domain/entity/LessonHistoryComment.kt
package com.tobe.healthy.lessonhistory.domain.entity

import com.tobe.healthy.common.BaseTimeEntity
import com.tobe.healthy.member.domain.entity.Member
import jakarta.persistence.*
import jakarta.persistence.CascadeType.ALL
import jakarta.persistence.FetchType.LAZY
import jakarta.persistence.GenerationType.IDENTITY
import lombok.ToString
import org.hibernate.annotations.ColumnDefault
import org.hibernate.annotations.DynamicUpdate
import kotlin.jvm.Transient

@Entity
@DynamicUpdate
@ToString
class LessonHistoryComment(

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "parent_id")
    @ToString.Exclude
    val parent: LessonHistoryComment? = null,

    val order: Int,

    var content: String,

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "writer_id")
    @ToString.Exclude
    val writer: Member? = null,

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "lesson_history_id")
    @ToString.Exclude
    val lessonHistory: LessonHistory? = null,

    @OneToMany(fetch = LAZY, mappedBy = "lessonHistoryComment", cascade = [ALL])
    @ToString.Exclude
    var files: MutableList<LessonHistoryFiles> = mutableListOf(),

    @Id
    @GeneratedValue(strategy = IDENTITY)
    @Column(name = "lesson_history_comment_id")
    val id: Long? = null,

    @ColumnDefault("false")
    var delYn: Boolean = false,

    @Transient
    var replies: MutableList<LessonHistoryComment> = mutableListOf(),

    ) : BaseTimeEntity<LessonHistoryComment, Long>() {

    fun updateLessonHistoryComment(content: String) {
        this.content = content
    }

    fun deleteComment() {
        this.delYn = true
    }
}
